packages/client-sdk-nodejs/src/internal/grpc/idle-grpc-client-wrapper.ts

```diff
@@ -48,31 +53,56 @@ export class IdleGrpcClientWrapper<T extends CloseableGrpcClient>
   }
 
   getClient(): T {
-    this.logger.trace(
-      `Checking to see if client has been idle for more than ${this.maxIdleMillis} ms`
-    );
+    let shouldRecreate = false;
+    let reason: string | undefined;
+
+    // Reconnect if channel is in a bad state.
+    // Although the generic type `T` only extends `CloseableGrpcClient` (which doesn't define `getChannel()`),
+    // we know that in practice, the client returned by `clientFactoryFn()` is a gRPC client that inherits from
+    // `grpc.Client`: https://grpc.github.io/grpc/node/grpc.Client.html
+    //
+    // Since `grpc.Client` defines a public `getChannel()` method, we can safely assume it's present on the client.
+    // To make TypeScript accept this, we cast the client to `unknown` first and then to `GrpcClientWithChannel`.
+    // This double-cast is necessary to override structural type checking and is safe in our controlled use case.
+    const clientWithChannel = this.client as unknown as GrpcClientWithChannel;
+    const channel = clientWithChannel.getChannel?.();
+    if (channel) {
+      const state = channel.getConnectivityState(true);
+      if (
+        state === ConnectivityState.TRANSIENT_FAILURE ||
+        state === ConnectivityState.SHUTDOWN
+      ) {
+        reason = `gRPC channel is in bad state: ${
+          state === ConnectivityState.TRANSIENT_FAILURE
+            ? 'TRANSIENT_FAILURE'
+            : 'SHUTDOWN'
+        }`;
+        shouldRecreate = true;
+      }
+    }
+
+    // Idle timeout check
     if (Date.now() - this.lastAccessTime > this.maxIdleMillis) {
-      this.logger.info(
-        `Client has been idle for more than ${this.maxIdleMillis} ms; reconnecting.`
-      );
-      this.client.close();
-      this.client = this.clientFactoryFn();
+      reason = `Client has been idle for more than ${this.maxIdleMillis} ms`;
+      shouldRecreate = true;
     }
 
+    // Max client age check
     if (this.maxClientAgeMillis !== undefined) {
-      this.logger.trace(
-        `Checking to see if client was created more than ${this.maxClientAgeMillis} ms`
-      );
       if (Date.now() - this.clientCreatedTime > this.maxClientAgeMillis) {
-        this.logger.info(
-          `Client was created more than ${this.maxClientAgeMillis} millis ago; recreating as asked.`
-        );
-        this.client.close();
-        this.client = this.clientFactoryFn();
-        this.clientCreatedTime = Date.now();
+        reason = `Client has been alive for more than ${this.maxClientAgeMillis} ms`;
+        shouldRecreate = true;
       }
     }
 
+    // Recreate the client if needed
+    if (shouldRecreate) {
+      this.logger.info(`${reason as string}; reconnecting client.`);
+      this.client.close();
+      this.client = this.clientFactoryFn();
+      this.clientCreatedTime = Date.now();
+    }
```

Comment:
I think this would be more readable if this logic was in a function called somethig like `recreateClient(reason: string)` that returned a new client. You could then just call that method as soon as you detect the client needs to be reconstructed. So everywhere you currently have `shouldRecreate = true` can be replaced with 

```
this.client = recreateClient("reason text here");
retrurn;
```

That way you wouldn't need to track the `shouldRecreate` boolean or consider cases that aren't relevant.